libkbfs/file_data.go

```diff
@@ -1523,37 +1835,84 @@ func (fd *fileData) deepCopy(ctx context.Context, codec kbfscodec.Codec,
 // undupChildrenInCopy takes a top block that's been copied via
 // deepCopy(), and un-deduplicates all leaf children of the block.  It
 // adds all child blocks to the provided `bps`, including both the
-// ones that were deduplicated and the ones that weren't (TODO).  It
-// returns the BlockInfos for all children, including ones that
-// weren't de-duplicated.
+// ones that were deduplicated and the ones that weren't.  It returns
+// the BlockInfos for all children.
 func (fd *fileData) undupChildrenInCopy(ctx context.Context,
 	bcache BlockCache, bops BlockOps, bps *blockPutState,
 	topBlock *FileBlock) ([]BlockInfo, error) {
 	if !topBlock.IsInd {
 		return nil, nil
 	}
 
-	// TODO: support multiple levels of indirection.
-	// TODO: parallelize
-	blockInfos := make([]BlockInfo, len(topBlock.IPtrs))
-	for i, iptr := range topBlock.IPtrs {
-		if iptr.RefNonce == kbfsblock.ZeroRefNonce {
-			// No need to un-dup mid-level indirect blocks.  TODO:
-			// fetch the blocks and add them to the bps (only needed
-			// for multiple levels of indirection).
-			blockInfos = append(blockInfos, iptr.BlockInfo)
-			continue
-		}
-		childBlock, _, err := fd.getter(ctx, fd.kmd, iptr.BlockPointer,
-			fd.file, blockRead)
+	// For indirect files, get all the paths to leaf blocks.  Note
+	// that because topBlock is a deepCopy, all of the blocks are also
+	// deepCopys and thus are modifiable.
+	pfr, err := fd.getIndirectBlocksForOffsetRange(ctx, topBlock, 0, -1)
+	if err != nil {
+		return nil, err
+	}
+	if len(pfr) == 0 {
+		return nil, fmt.Errorf(
+			"Indirect file %v had no indirect blocks", fd.rootBlockPointer())
+	}
+
+	// Append the leaf block to each path, since readyHelper expects
+	// it.
+	for i, path := range pfr {
+		leafPtr := path[len(path)-1].childIPtr().BlockPointer
+		leafBlock, _, err := fd.getter(
+			ctx, fd.kmd, leafPtr, fd.file, blockWrite)
```

Comment:
You're right, this could be a problem when multiple devices write to the same very large file at the same time, and the one that lost the race has journaling enabled.  To fix this, the best plan I can think of would be to have a `DirtyBlockCache` that is backed by the disk, and maybe a separate `ReadyBlockCache` that is also backed by disk (and then `blockPutState` would change to have pointers into the `ReadyBlockState`, rather than itself containing block data).  Paging this function won't really help, since all the blocks would still be put into those in-memory caches right now until the resolution is complete.

So, I guess my proposal is to go forward with this PR, and even the next one that enables large files in production.  But, we wait to turn on journaling by default until the disk cache is ready and we've fixed this code to use it.  And I'll add a TODO here, and warm internal journaling testers to be careful about writing to big files from multiple devices.  What do you think?